### DISCUSSION TITLE
Auto Backups

### DISCUSSION THREAD

**Discussion by sonicityV** [enrique-lozano/Monekin#257] [Ideas / New features]
Problem to solve

Currently the only way to create a backup is for the user to manually start the process. 

Possible solutions

I think the best thing would be to add a option to automatically create a backup file, options could include: "on change", "daily", "weekly" or "monthly". A choice on where to store the backup (instead just putting it downloads) would also be helpful.

**Comment by enrique-lozano** [comment]
This is a great idea that i've been thinking of for a while. 

I would like to implement the Google Drive backup in the future, but this sounds a good alternative to me for the moment. Both alternatives can even live together since some people may prefer to store the backup locally (without the need of an internet connection).

One minor thing to point out is that I would replace the "on change" option for the "on app exit" option, merely for performance reasons.

What do you think?

**Reply by sonicityV** [reply]
I reckon that would be pretty good, honestly I don't personally see a huge reason for "on change" or "on app exit" but, I get that people have different needs and I think "on app exit" would work.

Also slightly off topic, but wow I just started using this app and its so good 😄

**Reply by enrique-lozano** [reply]
Thank you very much! Comments from people like you are what encourage me to keep updating and improving it 😊

I created it because none of the thousands of personal finance apps met all my needs, and I honestly think that this it's the one that offers the most functionality in a simple and free way

**Reply by sonicityV** [reply]
Oh um out of curiosity, idk if I should have this as another feature request but is there any other way to donate apart from google play?

**Reply by enrique-lozano** [reply]
Yes and many thanks for asking that! You can reach to me via PayPal [here](https://paypal.me/lozanoelc?country.x=ES&locale.x=es_ES). I will upgrade the main GitHub page with this info also.

Feel free to open a new post about it if you want, for other users to now this possibility. It will not be available in the official app in Play Store since apps that allows other ways of payments have problems to get published.

**Reply by victor-marino** [reply]
Absolutely agree with this.

Manual backups are usually a bad solution, because:
1. Making manual daily backups is both cumbersome and easy to forget
2. Phones break or get lost unexpectedly, and when it happens, you probably haven't made a backup in 2 months and you'll lose lots of data 😅

Having some kind of automated solution is super important, bonus points if it allows you to save the backups to cloud storage (Google Drive, OneDrive, etc.) so that you don't lose them when you lose your phone.

**Reply by enrique-lozano** [reply]
Agree that cloud backups would be ideal, but this functionality is complicated and complex to implement and right now I don't have time to do it. As always, I appreciate PRs and collaborations in this regard.

In the meantime, I think that having local auto-backups would be a great addition and would not conflict with the cloud auto-backups that would be developed in the future.

**Reply by victor-marino** [reply]
Already replied to this in the other conversation, sorry for opening so many threads. I couldn't sleep last night 😅

As said I'll try to take a look at possible implementations, but can't promise anything as I'm really short on time these days.

Thanks a lot for your awesome app and for being so responsive to feedback! Not always the case with open source projects.

**Comment by victor-marino** [comment]
Hi @enrique-lozano.

I think a previous step towards implementing the auto-backup feature should be to first offer the user the option to select the folder where backups should be saved. Do you agree?

If so, I've been playing around with it and I think I have the logic more or less implemented (still need to tidy up the UI).

Should I open an issue for this to track any future PRs?

Also, one thing we'd need to do is store the preferred save path so that it persists between app restarts. I assume the right place for this should be a new `SettingKey` in the database, correct?

**Reply by enrique-lozano** [reply]
Hi @victor-marino! Thanks a lot for working in this. I will create an issue parting of this today if I can, so don't worry about that :)

I agree that starting by letting the user define the folder where to make backups would be a good starting point. However, I think everything should be done more or less at the same time.

Defining a good UI for this case seems to me perhaps one of the most complex tasks. I think we can look at [Bluecoins](https://play.google.com/store/apps/details?id=com.rammigsoftware.bluecoins) here, which offers exactly the same functionalities that we want to implement sooner or later. It has a button to download a backup on the fly (which we already have), another menu to configure automatic local backups (what we want now) and another menu to configure cloud backups (what we want in the future). I will put all this information in your corresponding issue so that it is better reflected.

Finally, yes, we must persist this information as you have mentioned in the settings table. We will have to add a migration file to a new version of the DB (v8) as well.

**Reply by victor-marino** [reply]
Agreed 👍🏻

I had initially added a simple text field at the bottom of the backup screen, which opens the file picker (to select a folder) when tapped:

![Screenshot_1734281070](https://github.com/user-attachments/assets/1e056bf5-291a-44fe-b598-1223991748b5)

But you're right that we need to think carefully how all the backup options will fit together in the UI first. I'll give it some thought.

One question: do you think it would make sense to include the option to export transaction data only (CSV) in the automatic backups (local or cloud)?

The way I see it, I think the CSV exports are probably a one-off thing that you would typically do manually from time to time, right? Automatic backups should probably be limited to the full database only.

Let me know if you agree.

**Reply by enrique-lozano** [reply]
Yep, agree, the CSV backups should remain outside of this new features. As you said, CSV exports are something that you would typically do manually from time to time.

Nice work with the select a folder picker by the way, keep investigating and we'll see what we get :)

**Comment by enrique-lozano** [comment]
Issue created in #268 with the requirements for this new feature 😊
